### ISSUE TITLE
DbContext.Entry().Reference().LoadAsync() considers added entity as Detached

### ISSUE THREAD

**Issue by Dennis-Petrov** [dotnet/efcore#19652]
When detached entity is added to unchanged entity collection, `LoadAsync` still considers added entity as `Detached`. This happens until parent (unchanged) entity state is not queried.

### Steps to reproduce

There's `Document` entity, which is `Unchanged` and has a `ComboStatus` collection.
This code adds new `comboStatus` into collection, and then tries to load `Event` reference:

```
    document.ComboStatuses.Add(comboStatus);

    await dbContext.Entry(comboStatus).Reference(_ => _.Event)
        .LoadAsync(cancellationToken)
        .ConfigureAwait(false);
```
`LoadAsync` fails with:

> Navigation property 'Event' on entity of type 'ComboStatus' cannot be loaded because the entity is > not being tracked. Navigation properties can only be loaded for tracked entities.

This is what I see in debugger:

![image](https://user-images.githubusercontent.com/5005677/72790263-5df5f980-3c46-11ea-946d-0f4650ff74cf.png)

And this is after querying `document`'s state:

![image](https://user-images.githubusercontent.com/5005677/72790389-8da50180-3c46-11ea-89ee-f74a21d396b9.png)

The stack trace:

>    at Microsoft.EntityFrameworkCore.Internal.EntityFinder`1.<LoadAsync>d__11.MoveNext()
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
>    at Im.ComboStatusTimer.<SaveComboStatusesAsync>d__23.MoveNext() in 
> 
### Further technical details

This is migration issue.
The code works in EF Core 2.2.

EF Core version: 3.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1

**Comment by ajcvickers** [comment]
@Dennis-Petrov Please attach a small, runnable project or post a complete code listing that reproduces the behavior you are seeing so that we can fully investigate.

**Comment by Dennis-Petrov** [comment]
@ajcvickers, I can't post complete code listing, it's a huge app with tons of code. I'll try to reproduce issue in a sample later. At the moment the workaround is to set entity state explicitly:

````
                        document.ComboStatuses.Add(comboStatus);

                        var entry = dbContext.Entry(comboStatus);

                        // this marks comboStatus as added
                        entry.State = EntityState.Added;

                        await entry.Reference(_ => _.Event)
                            .LoadAsync(cancellationToken)
                            .ConfigureAwait(false);

```

**Comment by Dennis-Petrov** [comment]
@ajcvickers , sample to reproduce error:
https://github.com/Dennis-Petrov/EF-Core-19652

**Comment by ajcvickers** [comment]
The issue here is that the local DetectChanges that happens when `Entry` is called does not detect the `comboStatus` entity added to the graph. This is a very similar situation to #17828.

@Dennis-Petrov Adding a call to `DetectChanges` should fix the issue:

```C#
document.ComboStatuses.Add(comboStatus);

context.ChangeTracker.DetectChanges();
                
await context.Entry(comboStatus)
    .Reference(_ => _.Event)
    .LoadAsync();
```

**Comment by Dennis-Petrov** [comment]
@ajcvickers , in my real code I add new items to the collection inside `foreach` loop, so, I think, that setting entity state explicitly will be faster, than call to `DetectChanges`, because the later must inspect entire graph (please correct me, if I wrong).

Should this issue be closed as duplicate of #17828?

**Comment by ajcvickers** [comment]
@Dennis-Petrov Yes, Add should be faster. Also, it was pointed out in our team discussion that setting a back-reference (if one exists) should also work:
```C#
comboStatus.Document = document;
```
This should work because then local DetectChanges on `comboStatus` will detect the change.
